### ISSUE TITLE
jit much slower with pytorch 1.5 (on custom rnns)

### ISSUE THREAD

**Issue by mravanelli** [pytorch/pytorch#38342]
## 🐛 Bug

During the development of SpeechBrain, we observed that our jitted custom rnn models are significantly slower with pytorch 1.5 compared to pytorch 1.4. After doing some tests, I realized that any kind of "long" for loop (even doing simple operations only) is much slower. Let me share a minimal example in the following (I have a larger example that I can share on demand) that basically implements a simple counter. I noticed the same issues in different clusters and machines.

## To Reproduce

```
import torch
import torch.nn as nn
import time


class custom_rnn(torch.jit.ScriptModule):

    def __init__(
        self,
        input_size,
        hidden_size,
        num_layers,
        batch_size,
        nonlinearity="relu",
        device="cuda",
    ):

        super(custom_rnn, self).__init__()
        self.hidden_size = int(hidden_size)
        self.input_size = int(input_size)
        self.batch_size = batch_size
        self.device = device

        self.w = nn.Linear(
            self.input_size, 2 * self.hidden_size, bias=False
        ).to(device)
        
        # Initilizing initial state h
        self.h_init = torch.zeros(
                self.batch_size,
                self.hidden_size,
                requires_grad=False,
                device=self.device,
            )

    @torch.jit.script_method
    def forward(self, x):
        ht = self.h_init
        
        # Loop over time axis
        for k in range(x.shape[1]):
            ht = ht + 1.0
        return ht

    
if __name__ == "__main__":

    inp_tensor = torch.rand([4, 500, 40]).to('cuda')
    net = custom_rnn(40, 512, 1, 4, device='cuda').to('cuda')
    start = time.time()

    for i in range(1000):
        out_tensor = net(inp_tensor)

    end = time.time()
    print(end - start)
 
```

## Expected behavior

Pytorch 1.4 takes on average 0.85 sec, while PyTorch 1.5 takes 5.49 sec(!!). Another example (not reported here) based on full custom rnn shows similar speed issues.

## Environment

I'm using pytorch 1.5 and 1.4 (both stable versions) installed with conda.

Any idea? Thank you in advance for your help!

cc @suo

**Comment by suo** [comment]
cc @ZolotukhinM, @Krovatkin

**Comment by mravanelli** [comment]
Dear @suo @ZolotukhinM, are you able to replicate my speed issue?

**Comment by mravanelli** [comment]
Hi @Krovatkin,
what do you mean with PE, LE, SE?

**Comment by Krovatkin** [comment]
@mravanelli these are different execution modes of PyTorch JIT. 
LE (legacy executor) was the default mode in 1.4. It uses the old fuser infrastructure and it was a bit slow to compile in some instances.
SE (simple executor) was the default mode in 1.5 It has no fusion, very fast compilation times, but not ideal for our users who depend on fusion.
PE was the default on master before 1.5 and this is the mode we are working on improving.

WIth have identified the issue and will be working on fixing this, soon hopefully!

**Comment by Krovatkin** [comment]
@mravanelli I'm sorry I was mostly documenting the issue for myself))))

**Comment by vincentqb** [comment]
@Krovatkin I believe there is a way to change the executor if needed. Can you point to a reference on how to do this?

**Comment by Krovatkin** [comment]
@vincentqb  
```python
        torch._C._jit_set_profiling_executor(False)
        torch._C._jit_set_profiling_mode(False)
```

Also, please see for more details:

https://github.com/pytorch/pytorch/blob/master/benchmarks/fastrnns/bench.py#L212-L241

**Comment by vincentqb** [comment]
Thanks @mravanelli for giving the following offline update on this.

Without the flags mentioned by @Krovatkin in [comment](https://github.com/pytorch/pytorch/issues/38342#issuecomment-644324034):
- pytorch 1.7 takes about 6.3 seconds
- pytorch 1.6 takes 5.49 seconds
- pytorch 1.5 takes 0.85 seconds

With flags, all three takes about 0.85 seconds.

@Krovatkin -- do you have an update on this? thanks!

**Comment by ZolotukhinM** [comment]
I've just built pytorch 1.7 from sources and re-measured the numbers, here is what I got:
* default settings: **0.89s**
* PE turned on explicitly (should be the same as the default): **0.84s**
```
    torch._C._jit_set_profiling_executor(True)
    torch._C._jit_set_profiling_mode(True)
```
* LE (should roughly match 1.4 and 1.6 behavior): **1.36s**
```
    torch._C._jit_set_profiling_executor(False)
    torch._C._jit_set_profiling_mode(False)
```
* SE (should match 1.5 behavior): **7.45s**
```
    torch._C._jit_set_profiling_executor(True)
    torch._C._jit_set_profiling_mode(False)
```

@vincentqb could you please run the same measurements on 1.7? I'm puzzled by how different our measured numbers seem to be.

**Comment by vincentqb** [comment]
Thanks for double checking @ZolotukhinM :) The numbers were quoted from @mravanelli offline, do you want to comment?

I've rerun them, and I can't reproduce. 
```
DEFAULT
0.9890170097351074
PE
0.8147268295288086
LE
0.8105366230010986
SE
0.8073666095733643

Version:  1.7.0+cu101
```

<details><summary>script</summary>

```python
import torch
import torch.nn as nn
import time


class custom_rnn(torch.jit.ScriptModule):

    def __init__(
        self,
        input_size,
        hidden_size,
        num_layers,
        batch_size,
        nonlinearity="relu",
        device="cuda",
    ):

        super(custom_rnn, self).__init__()
        self.hidden_size = int(hidden_size)
        self.input_size = int(input_size)
        self.batch_size = batch_size
        self.device = device

        self.w = nn.Linear(
            self.input_size, 2 * self.hidden_size, bias=False
        ).to(device)
        
        # Initilizing initial state h
        self.h_init = torch.zeros(
                self.batch_size,
                self.hidden_size,
                requires_grad=False,
                device=self.device,
            )

    @torch.jit.script_method
    def forward(self, x):
        ht = self.h_init
        
        # Loop over time axis
        for k in range(x.shape[1]):
            ht = ht + 1.0
        return ht

def run():

    inp_tensor = torch.rand([4, 500, 40]).to('cuda')
    net = custom_rnn(40, 512, 1, 4, device='cuda').to('cuda')
    start = time.time()

    for i in range(1000):
        out_tensor = net(inp_tensor)

    end = time.time()
    print(end - start)


print("DEFAULT")
run()

torch._C._jit_set_profiling_executor(True)
torch._C._jit_set_profiling_mode(True)

print("PE")
run()

torch._C._jit_set_profiling_executor(False)
torch._C._jit_set_profiling_mode(False)

print("LE")
run()

torch._C._jit_set_profiling_executor(True)
torch._C._jit_set_profiling_mode(False)

print("SE")
run()

print("Version: ", torch.__version__)
```

</details>

**Comment by mravanelli** [comment]
On my side I noticed that our customized and jitted RNN with pytorch 1.7 is
15% slower than the same RNN with pytorch 1.6.
When adding the following flags:

torch._C._jit_set_profiling_executor(True)
torch._C._jit_set_profiling_mode(True)

Everything runs fast.

**Comment by ZolotukhinM** [comment]
@vincentqb , in your script you're changing executor setting in between function invocations - I don't think it actually works this way. We create an executor for the function the first time it was called and then use it for that function no matter how you change those settings afterwards (I admit that this behavior could seem surprising, but that's an internal API after all). A better experiment would be to set these setting once in the script and run the script several times (changing the settings in the script between the runs).

**Comment by Krovatkin** [comment]
cc @vincentqb : 

I couldn't reproduce the regression either.
@ZolotukhinM , I think something like this could work (although I need to test it thoroughly?

1) We do `torch.jit.load` to get a brand new model
2) We do `warmup` to make PE and LE compile the optimized versions of the model.l


```python
import torch
import torch.nn as nn
import time


class custom_rnn(torch.jit.ScriptModule):

    def __init__(
        self,
        input_size,
        hidden_size,
        num_layers,
        batch_size,
        nonlinearity="relu",
        device="cuda",
    ):

        super(custom_rnn, self).__init__()
        self.hidden_size = int(hidden_size)
        self.input_size = int(input_size)
        self.batch_size = batch_size
        self.device = device

        self.w = nn.Linear(
            self.input_size, 2 * self.hidden_size, bias=False
        ).to(device)
        
        # Initilizing initial state h
        self.h_init = torch.zeros(
                self.batch_size,
                self.hidden_size,
                requires_grad=False,
                device=self.device,
            )

    @torch.jit.script_method
    def forward(self, x):
        ht = self.h_init
        
        # Loop over time axis
        for k in range(x.shape[1]):
            ht = ht + 1.0
        return ht

def get_fresh():
    return torch.jit.load('scriptmodule.pt')


def run(net, inp_tensor):
    

    start = time.time()

    
    for i in range(1000):
        out_tensor = net(inp_tensor)

    end = time.time()
    print(end - start)


def warmup(net, inp_tensor):
    for i in range(20):
        out_tensor = net(inp_tensor)

net = custom_rnn(40, 512, 1, 4, device='cuda').to('cuda')
torch.jit.save(net, 'scriptmodule.pt')
inp_tensor = torch.rand([4, 500, 40]).to('cuda')

print("DEFAULT")
net = get_fresh()
warmup(net, inp_tensor)
run(net, inp_tensor)
torch._C._jit_set_profiling_executor(True)
torch._C._jit_set_profiling_mode(True)

print("PE")
net = get_fresh();
warmup(net, inp_tensor)
run(net, inp_tensor)

torch._C._jit_set_profiling_executor(False)
torch._C._jit_set_profiling_mode(False)

print("LE")
net = get_fresh();
warmup(net, inp_tensor)
run(net, inp_tensor)

torch._C._jit_set_profiling_executor(True)
torch._C._jit_set_profiling_mode(False)

print("SE")
net = get_fresh();
warmup(net, inp_tensor)
run(net, inp_tensor)

print("Version: ", torch.__version__)

```

**Comment by vincentqb** [comment]
Thanks for the clarifications!

Running the script 4 times as mentioned [above](https://github.com/pytorch/pytorch/issues/38342#issuecomment-724399102) give:

```
❯ python script1.py
DEFAULT
1.0017213821411133
Version:  1.7.0+cu101

❯ python script2.py
PE
0.9959783554077148
Version:  1.7.0+cu101

❯ python script3.py
LE
1.3876075744628906
Version:  1.7.0+cu101

❯ python script4.py
SE
7.5840888023376465
Version:  1.7.0+cu101
```

And thanks for [the suggestion](https://github.com/pytorch/pytorch/issues/38342#issuecomment-724878921), I get
```
DEFAULT
0.843116283416748
PE
0.8396866321563721
LE
1.0090396404266357
SE
8.275022029876709
Version:  1.7.0+cu101
```

**Comment by ZolotukhinM** [comment]
It looks like the issue could be closed now, please reopen if you notice other performance regressions and we will be happy to look into them!
